vault/expiration_test.go

```diff
@@ -424,6 +424,73 @@ func TestExpiration_TotalLeaseCount_WithRoles(t *testing.T) {
 	}
 }
 
+// TestExpiration_EmitsEvents tests that lease creation and revocation emits events
+func TestExpiration_EmitsEvents(t *testing.T) {
+	// Quotas and internal lease count tracker are coupled, so this is a proxy
+	// for testing the total lease count quota
+	c, _, _ := TestCoreUnsealed(t)
+
+	exp := c.expiration
+	root, err := exp.tokenStore.rootToken(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// count events
+	events, cancel, err := c.Events().Subscribe(context.Background(), namespace.RootNamespace, "core/lease-*", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cancel()
+	eventsCount := atomic.Int32{}
+
+	timeout := time.NewTimer(30 * time.Second)
+	defer timeout.Stop()
+	go func() {
+		for {
+			select {
+			case <-timeout.C:
+				return
+			case <-events:
+				eventsCount.Add(1)
+			}
+		}
+	}()
+
+	auth := &logical.Auth{
+		ClientToken: root.ID,
+		LeaseOptions: logical.LeaseOptions{
+			TTL: 10 * time.Millisecond,
+		},
+	}
+
+	te := &logical.TokenEntry{
+		Path:        "auth/github/login",
+		NamespaceID: namespace.RootNamespaceID,
+	}
+	err = exp.RegisterAuth(namespace.RootContext(nil), te, auth, "")
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// lease create + revoke events
+	expectedEvents := int32(2)
+	check := time.NewTicker(100 * time.Millisecond)
+	defer check.Stop()
+	for {
+		select {
+		case <-timeout.C:
+			if expectedEvents != eventsCount.Load() {
```

Comment:
Would this be better off checking for event types, to prevent the test breaking if we ever start emitting events for other reasons, and to ensure that we're not just getting say two createlease events instead of one create and one expire?

Reply:
Makes sense. I was trying to keep it simple, but you're right, we should check specific types.